source/io/module_javaScriptInvoke.js

```diff
@@ -153,198 +177,339 @@
+            javaScriptReturnTypeName,
+            returnValuePointer,
+            returnValue) {
+            if (!(returnValue instanceof Uint8Array)) {
+                throwTypeMismatchException();
+            }
+
+            Module.eggShell.dataWriteUInt8Array(returnValuePointer, returnValue);
+        };
+
+        var updateUInt16ArrayReturnValue = function (
+            javaScriptReturnTypeName,
+            returnValuePointer,
+            returnValue) {
+            if (!(returnValue instanceof Uint16Array)) {
+                throwTypeMismatchException();
             }
+
+            Module.eggShell.dataWriteUInt16Array(returnValuePointer, returnValue);
+        };
+
+        var updateUInt32ArrayReturnValue = function (
+            javaScriptReturnTypeName,
+            returnValuePointer,
+            returnValue) {
+            if (!(returnValue instanceof Uint32Array)) {
+                throwTypeMismatchException();
+            }
+
+            Module.eggShell.dataWriteUInt32Array(returnValuePointer, returnValue);
+        };
+
+        var updateSingleArrayReturnValue = function (
+            javaScriptReturnTypeName,
+            returnValuePointer,
+            returnValue) {
+            if (!(returnValue instanceof Float32Array)) {
+                throwTypeMismatchException();
+            }
+
+            Module.eggShell.dataWriteSingleArray(returnValuePointer, returnValue);
+        };
+
+        var updateDoubleArrayReturnValue = function (
+            javaScriptReturnTypeName,
+            returnValuePointer,
+            returnValue) {
+            if (!(returnValue instanceof Float64Array)) {
+                throwTypeMismatchException();
+            }
+
+            Module.eggShell.dataWriteDoubleArray(returnValuePointer, returnValue);
+        };
+
+        var isTypedArray = function (
+            value) {
+            if (value instanceof Int8Array ||
+                value instanceof Int16Array ||
+                value instanceof Int32Array ||
+                value instanceof Uint8Array ||
+                value instanceof Uint16Array ||
+                value instanceof Uint32Array ||
+                value instanceof Float32Array ||
+                value instanceof Float64Array) {
+                return true;
+            }
+
+            return false;
+        };
+
+        var isValidJavaScriptReturnType = function (
+            returnValue) {
+            var returnTypeName = typeof returnValue;
+            return (returnTypeName === 'number') ||
+            (returnTypeName === 'boolean') ||
+            (returnTypeName === 'string') ||
+            (returnTypeName === 'undefined') ||
+            (isTypedArray(returnValue));
+        };
+
+        var updateReturnValueGivenType = function (
+            returnTypeName,
+            returnValuePointer,
+            returnValue) {
+            var javaScriptReturnTypeName = typeof returnValue;
             switch (returnTypeName) {
```

Comment:
I'm concerned about these switch statements that keep growing. Perhaps we can do something about them before is too late to refactor.
Consider having an object that has all the information of which functions to call depending on the type.
If we could have an structure similar to:
```javascript
var types_io = { // Please choose a better name.
	'Int8': {
		reader: Module.eggShell.dataReadInt8,
		writer: Module.eggShell.dataWriteInt8,
		isValidReturnType : isJavaScriptNumber
	}, ..., 
	'ArrayDouble' : {
		reader: Module.eggShell.dataReadDoubleArray,
		writer: Module.eggShell.dataWriteDoubleArray,
		isValidReturnType : isFloat64Array
	}
}
```

Then we could get rid of both switches that we have. As an example, here is how the `updateReturnValueGivenType` would look like.
```javascript
var updateReturnValueGivenType = function (
            returnTypeName,
            returnValuePointer,
            returnValue) {
	var javaScriptReturnTypeName = typeof returnValue;
	var typeConfig = types_io[returnTypeName]; // If this is undefined then the type is not registered.
	
	if (typeConfig.isValidReturnType(returnValue) === false) {
		// Not sure if we should throw here or just let the callee know that it failed.
	}
	
	var writer = typeConfig.writer(returnValuePointer, returnValue);
}
```

Something similar would happen with `createJavaScriptParametersArray()`. Also we could get rid of all the `update*ReturnValue()` family of functions since they only validate the value type and then call `dataWrite*()`. Here are the implementations for `isValidReturnType`:

```javascript
var isJavaScriptNumber = function (value) {
	return typeof value === 'number';
};

var isFloat64Array = function (value) {
	return value instanceof Float64Array;
};
```

Reply:
Last line of `updateReturnValueGivenType` is wrong.
```javascript
typeConfig.writer(returnValuePointer, returnValue);
```

Reply:
Great suggestion.

Reply:
I like it. I needed to refactor this method to begin with to avoid a complexity linting rule, and this will make maintaining that requirement even easier.